### ISSUE TITLE
Specify concatenation `axis` to `apply_ragged`

### ISSUE THREAD

**Issue by milancurcic** [Cloud-Drift/clouddrift#274]

**Comment by milancurcic** [comment]
I started looking into this. Specifying `axis` to `apply_ragged` requires:

1. Specifying `axis` along which we unpack at the beginning of `apply_ragged`;
2. Specifying `axis` along which we concatenate at the end of `apply_ragged`.

2 is easy--we just specify `axis` to `np.concatenate`. So, let's discuss 1, which is a bit more complicated.

In step 1, we unpack the ragged array using `unpack_ragged`. Currently, `unpack_ragged` always unpacks along `axis=0`. Rather than attempting to do `np.moveaxis` acrobatics inside of `unpack_ragged`, I first looked whether NumPy provides a function that's the inverse of `np.concatenate` and that allows specifying `axis`. It turns out it does and it's called `np.split`:

https://numpy.org/doc/stable/reference/generated/numpy.split.html

The second argument to `np.split` is `indices_or_sections`, which in our case would be `rowsize_to_index(rowsize)[1:-1]`. The result is the same as that of `unpack_ragged`, so, we should really just rely on `np.split` and not have `unpack_ragged` as part of our API.

One downside to `np.split` is that it doesn't support choosing rows, introduced in #272.

Thus:

| Function | `axis` | `rows` | Source |
|----------|----------|----------|--------|
| `unpack_ragged`   | No   | Yes   | clouddrift |
| `np.split`  | Yes   | No   | NumPy |


We have a few options going forward:

1. Expand `unpack_ragged` to support selecting the unpacking `axis`
2. Remove `unpack_ragged` from CloudDrift and use `np.split`; in `apply_ragged` select `rows` after the ragged array has been completely unpacked by `np.split`.

Both `np.split` and `unpack_ragged` return _views_ into the source array and don't make a copy. However, there may be some performance advantage of `unpack_ragged` with internal `rows` where only some and not all slices are put into a list. I recommend not using the possibility of this performance advantage as a strong argument in favor of `unpack_ragged`.

What do you think? Currently I'm in favor of option 2, that is, removing `unpack_ragged` and using `np.split`.

**Comment by philippemiron** [comment]
I will also go with 2 and put the axis afterward in `np.concatenate` so the data that we pass goes directly to the function without having a implicit change of its shape.

With regards to `unpack_ragged` and `np.split`, can't we keep `unpack_ragged` and simply make this function wrap `np.split` to add that `row` capability? Since `np.split` returns a view, it should be very fast on the full array, and then we would just return a subset of the rows if needed.

**Comment by milancurcic** [comment]
> I will also go with 2 and put the axis afterward in np.concatenate so the data that we pass goes directly to the function without having a implicit change of its shape.

Yes, `np.concatenate` is not an issue. Adding `axis` to `np.concatenate` would be needed for both options 1 and 2.

> With regards to unpack_ragged and np.split, can't we keep unpack_ragged and simply make this function wrap np.split to add that row capability? Since np.split returns a view, it should be very fast on the full array, and then we would just return a subset of the rows if needed.

We can do that. That would be useful if we expect the unpacking-by-rows to be useful more broadly than just in `apply_ragged` (I think it would be). I like your suggested approach the best.

**Comment by selipot** [comment]
I like the idea of keeping `unpack_ragged` even as a wrapper for `np.split`. It makes sense for this function to exist within our library built around ragged arrays. As you wrote above for the user to use `np.split` requires the user to know that function and to build the arguments for it i.e `rowsize_to_index(rowsize)[1:-1]`.
